engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb

```diff
@@ -0,0 +1,169 @@
+require 'rails_helper'
+
+describe Api::Connect::V3::Systems::ProductsController, type: :request do
+  include_context 'auth header', :system, :login, :password
+  include_context 'version header', 3
+
+  let(:url) { connect_systems_products_url }
+  let(:headers) { auth_header.merge(version_header) }
+  let(:product) { FactoryGirl.create(:product, :with_mirrored_repositories, :with_mirrored_extensions) }
+
+  let(:payload) do
+    {
+      identifier: product.identifier,
+      version: product.version,
+      arch: product.arch
+    }
+  end
+
+  describe '#activate' do
+    context "when system doesn't have hw_info" do
+      let(:system) { FactoryGirl.create(:system) }
+
+      it 'class instance verification provider' do
+        expect(InstanceVerification::Providers::Example).to receive(:new)
+          .with(be_a(ActiveSupport::Logger), be_a(ActionDispatch::Request), payload, nil).and_call_original
+        post url, params: payload, headers: headers
+      end
+    end
+
+    context 'when system has hw_info' do
+      let(:instance_data) { 'dummy_instance_data' }
+      let(:system) { FactoryGirl.create(:system, :with_hw_info, instance_data: instance_data) }
+      let(:serialized_service_json) do
+        V3::ServiceSerializer.new(
+          product.service,
+          base_url: URI::HTTP.build({ scheme: response.request.scheme, host: response.request.host }).to_s
+        ).to_json
+      end
+
+      context 'when verification provider returns false' do
+        before do
+          expect(InstanceVerification::Providers::Example).to receive(:new)
+            .with(be_a(ActiveSupport::Logger), be_a(ActionDispatch::Request), payload, instance_data).and_return(plugin_double)
+          expect(plugin_double).to receive(:instance_valid?).and_return(false)
+          post url, params: payload, headers: headers
+        end
+
+        let(:plugin_double) { instance_double('InstanceVerification::Providers::Example') }
```

Comment:
nitpicking: Why not move it to the upper level context on remove the other definitions of it?

Reply:
Sure, fixed.